In [1]:
import numpy as np
import pandas as pd

pd.set_option('display.max_rows', 200)

In [2]:
data = pd.read_csv("data_with_gmap_features.csv")

print(data.shape)
data.head()

(5849, 10)


neighborhood    price  beds  baths  \
0         olaleye estate alaka iponri surulere lagos  2500000     3      4   
1  off kudirat abiola way, julie estateoregun ike...  2500000     3      3   
2                       ...lekki phase 1 lekki lagos   800000     1      1   
3                          lekki phase 1 lekki lagos  1200000     1      1   
4                                         ketu lagos  1000000     2      2   

                                               extra  state locality  \
0  brand new service 3 bedroom flat all rooms ens...  Lagos    Lagos   
1  the property is a 3 bedrooms upscale flat all ...  Lagos    Ikeja   
2  for more information, pricing, and private vie...  Lagos    Lekki   
3  for more information, pricing, and private vie...  Lagos    Lekki   
4  brand new 2 bedrooms flat for rent at ketu lag...  Lagos    Lagos   

        lga           hood  zipcode  
0  Surulere        Oke Ira      NaN  
1     Ikeja          Allen      NaN  
2       NaN  Lekki Phase 1      NaN  
3       NaN  Lekki Phase 1      NaN  
4    Kosofe           Ketu      NaN

## Checking the Missing Values

In [3]:
data.isnull().sum()

neighborhood       0
price              0
beds               0
baths              0
extra              0
state              0
locality         172
lga              697
hood             924
zipcode         5210
dtype: int64

## Dropping `ZipCode`

Most of the addresses don't have a zip code (5210 out of 5849) due to bad address format or something, so I'll drop zipcode.

In [4]:
data.drop('zipcode', 1, inplace=True)
data.head()

neighborhood    price  beds  baths  \
0         olaleye estate alaka iponri surulere lagos  2500000     3      4   
1  off kudirat abiola way, julie estateoregun ike...  2500000     3      3   
2                       ...lekki phase 1 lekki lagos   800000     1      1   
3                          lekki phase 1 lekki lagos  1200000     1      1   
4                                         ketu lagos  1000000     2      2   

                                               extra  state locality  \
0  brand new service 3 bedroom flat all rooms ens...  Lagos    Lagos   
1  the property is a 3 bedrooms upscale flat all ...  Lagos    Ikeja   
2  for more information, pricing, and private vie...  Lagos    Lekki   
3  for more information, pricing, and private vie...  Lagos    Lekki   
4  brand new 2 bedrooms flat for rent at ketu lag...  Lagos    Lagos   

        lga           hood  
0  Surulere        Oke Ira  
1     Ikeja          Allen  
2       NaN  Lekki Phase 1  
3       NaN  Lekki Phase 1  
4    Kosofe           Ketu

## Cleaning `Locality`

In [5]:
data.groupby('locality')[['neighborhood', 'state', 'lga', 'hood']].count()

neighborhood  state   lga  hood
locality                                                    
Abeokuta                               12     12    11     6
Abuja                                 689    689   659   591
Abẹ́òkúta                               3      3     3     0
Agbara                                  1      1     1     0
Apata Ibadan                            1      1     1     0
Arepo                                  11     11    11     6
Arepo Berger                            1      1     1     0
Badagry                                 3      3     3     3
Benin City                              3      3     3     3
Central                                 2      2     2     2
Command                                 1      1     1     1
Egbe                                    1      1     1     1
Epe                                     1      1     1     0
FHA                                     1      1     1     1
Gbagada Expressway                      1      1     1     1
Germantown                              1      1     1     1
Gwagwalada                              2      2     2     0
I                                       1      1     1     1
ISAWO                                   1      1     1     1
Ibadan                                466    466   400   220
Ibadan, Oyo                            10     10    10     2
Ibadan, Oyo State                       1      1     1     0
Ibafo                                   8      8     8     0
Igbesa                                  1      1     1     0
Ikeja                                 162    162   162   129
Ikeja, Lagos, Nigeria                   8      8     8     0
Ikorodu                                93     93    92    93
Ilishan-Remo                            1      1     1     0
Iperu                                   1      1     1     0
Iwo Road                                1      1     1     0
Karu L. G. A Nasarawa State             1      1     1     0
Lagos                                3083   3083  2713  2824
Lekki                                 651    651   522   620
Magboro                                52     52    51     0
New Karu                                2      2     2     2
New York                                1      1     1     1
Odo Ogun                                1      1     1     0
Ojata                                   1      1     1     1
Ojodu                                 127    127    95   119
Ojodu Berger                           33     33    14    15
Ojokoro                                 1      1     1     1
Ojurin                                  3      3     3     3
Orozo                                   1      1     1     0
Ota                                    11     11    11     1
Owode                                   7      7     7     0
Oworonshoki                           102    102   102   102
Ozuoba                                  2      2     2     0
PHC                                     1      1     1     1
Port Harcourt                          72     72    61    63
Porter                                  1      1     1     0
Sagamu                                  5      5     5     1
Sangotedo                              24     24    24    19
Street                                  1      1     1     1
Washington                              1      1     0     1
Ìbàdàn                                  5      5     3     0

In [6]:
data.loc[data.locality.str.contains('Abẹ́òkúta') == True, 'locality'] = 'Abeokuta'
data.loc[data.locality.str.contains('Ibadan|Ìbàdàn|Ojurin|Iwo Road|Odo Ogun', na=False), 'locality'] = 'Ibadan'
data.loc[data.locality.str.contains('Ikeja', na=False), 'locality'] = 'Ikeja'
data.loc[data.locality.str.contains('I|Street', na=False), 'locality'] = 'Lekki'
data.loc[data.locality.str.contains('Command|Egbe|Epe|Gbagada Expressway|Ojata|Ojokoro', na=False), 'locality'] = 'Lagos'
data.loc[data.locality.str.contains('Ozuoba|PHC', na=False), 'locality'] = 'PortHarcourt'
data.loc[data.locality.str.contains('Central|FHA|Gwagwalada|New Karu|Orozo', na=False), 'locality'] = 'Abuja'

In [7]:
locality_group = data.groupby('locality')[['neighborhood', 'state', 'lga', 'hood']].count().sort_values(by='neighborhood')
locality_group

neighborhood  state   lga  hood
locality                                                    
Washington                              1      1     0     1
Agbara                                  1      1     1     0
Arepo Berger                            1      1     1     0
Germantown                              1      1     1     1
Karu L. G. A Nasarawa State             1      1     1     0
Porter                                  1      1     1     0
New York                                1      1     1     1
Badagry                                 3      3     3     3
Benin City                              3      3     3     3
PortHarcourt                            3      3     3     1
Sagamu                                  5      5     5     1
Owode                                   7      7     7     0
Arepo                                  11     11    11     6
Ota                                    11     11    11     1
Abeokuta                               15     15    14     6
Sangotedo                              24     24    24    19
Ojodu Berger                           33     33    14    15
Magboro                                52     52    51     0
Port Harcourt                          72     72    61    63
Oworonshoki                           102    102   102   102
Ojodu                                 127    127    95   119
Abuja                                 697    697   667   596
Lekki                                1416   1416  1218  1070
Lagos                                3089   3089  2719  2829

In [8]:
locality_to_drop = locality_group[(locality_group.neighborhood < 3) | (locality_group.index.str.contains('Benin'))]
locality_to_drop

neighborhood  state  lga  hood
locality                                                   
Washington                              1      1    0     1
Agbara                                  1      1    1     0
Arepo Berger                            1      1    1     0
Germantown                              1      1    1     1
Karu L. G. A Nasarawa State             1      1    1     0
Porter                                  1      1    1     0
New York                                1      1    1     1
Benin City                              3      3    3     3

In [9]:
locality_to_drop_str = str(locality_to_drop.index.tolist())[1:-1].replace(", ", "|").replace("'", "")
locality_to_drop_str

'Washington|Agbara|Arepo Berger|Germantown|Karu L. G. A Nasarawa State|Porter|New York|Benin City'

In [10]:
locality_to_drop_df = data.loc[data.locality.str.fullmatch(locality_to_drop_str) == True]
locality_to_drop_df

neighborhood    price  beds  baths  \
1052                iyanera. agbaraokokomaiko ojo lagos   170000     1      1   
1761       hazmat street, macaulayigbogbo ikorodu lagos   180000     4      2   
4578                sharpcorner-mararaba.mararaba abuja   140000     1      1   
4931  3 bedroom apartment at abuja quarters g.r.a be...   450000     5      3   
4950  3 bedroom apartment at abuja quarters off powe...   450000     3      3   
4956                          around sentracojabi abuja  1600000     5      3   
4989  3 bedroom apartment at abuja quarters g.r.a be...   450000     5      3   
5635    private estate, off berger expresswayarepo ogun   550000     3      3   
5741          arepo in a private estatearepo arepo ogun  1000000     4      4   
5786  peace estate, behind first bank close to brain...   900000     1      3   

                                                  extra   state  \
1052  with close proximity to the lagos state univer...   Lagos   
1761  annual rent:-180,000 a&c:-60,000 total package...   Lagos   
4578  tastefully finished swlfxontain apartment for ...   Abuja   
4931  3 bedroom apartment at abuja quarters g.r.a be...   Abuja   
4950  3 bedroom apartment at abuja quarters off powe...   Abuja   
4956  it's a spacious 2 bedroom with all rooms ensui...   Abuja   
4989  3 bedroom apartment at abuja quarters g.r.a be...   Abuja   
5635  3 bedroom flat / apartment for rent private es...    Ogun   
5741  furnished newly builtlovely 4 bed room duplex ...    Ogun   
5786  newly built block of flats consisting 3 bedroo...  Rivers   

                         locality              lga        hood  
1052                       Agbara      Ado Odo/Ota         NaN  
1761                   Washington              NaN   Navy Yard  
4578  Karu L. G. A Nasarawa State             Karu         NaN  
4931                   Benin City            Oredo         Oka  
4950                   Benin City             Egor  Okhokhugbo  
4956                     New York  New York County   Manhattan  
4989                   Benin City            Oredo         Oka  
5635                 Arepo Berger    Obafemi Owode         NaN  
5741                   Germantown  Columbia County    Clermont  
5786                       Porter   Wagoner County         NaN

In [11]:
data = data.drop(locality_to_drop_df.index)
data.shape

(5839, 9)

### Dropping the listings without `Locality` and `LGA`

In [12]:
data[(data.locality.isna()) & (data.lga.isna())]

neighborhood    price  beds  baths  \
53    alagomeji-adekunle, close to the 3rd mainland ...  1500000     2      2   
217                     ilado areaigbogbo ikorodu lagos   350000     5      3   
229                    off happy land estate ajah lagos  1100000     2      2   
324        irawo owode onirinmile 12 kosofe ikosi lagos   600000     2      3   
759                                     -maryland lagos  2000000     3      3   
854             iyanera - ketu axisokokomaiko ojo lagos   200000     2      1   
1079                 iyanera - ketuokokomaiko ojo lagos   200000     3      3   
1267                                     alimosho lagos   350000     4      3   
2199       owode onirin irawomile 12 kosofe ikosi lagos   350000     5      2   
2459       irawo owode onirinmile 12 kosofe ikosi lagos   500000     5      2   
2466   ayobo, close to megida bus stopayobo ipaja lagos   150000     1      2   
2722   ayobo, close to megida bus stopayobo ipaja lagos   180000     1      3   
2981                             -shasha alimosho lagos   400000     1      1   
3048        irawo, ondo streetkosofe kosofe ikosi lagos   550000     2      2   
3294     nice and secure estateosapa london lekki lagos  2100000     2      2   
3393               close to happyland estate ajah lagos  1200000     2      3   
3482  assese, off lagos-ibadan expresswayobafemi owo...   200000     2      2   
3644              0d0zi street,0j0du,berger ojodu lagos   600000     1      2   
3651  agboyi road alapere ketualapere kosofe ikosi l...   250000     1      1   
3713  ajiboye street alapere towards estate bus stop...   800000     3      1   
3729  agboyi road alapere ketualapere kosofe ikosi l...   550000     4      2   
3838  osapa london estate by shoprite roadosapa lond...   950000     1      1   
3928  assese, off lagos-ibadan expresswayobafemi owo...   250000     2      2   
3945                       oludare estateado ajah lagos   650000     2      3   
4176               isheri northisheri north ojodu lagos   750000     2      2   
4298  agboyi road alapere ketualapere kosofe ikosi l...   400000     1      1   
4318                          0j0du,,berger ojodu lagos   800000     3      3   
5101  peace estate directly beside, behind boldplace...   600000     2      2   
5258  adamawa idi oya off akala expressakala express...   250000     1      1   
5382        salute off ak ala express ibadan ibadan oyo   300000     1      1   
5473  aba ado off akala express ibadan akala express...   150000     1      1   
5658  assese, off lagos-ibadan expresswayobafemi owo...   200000     2      2   
5672                                      nujarepo ogun   650000     3      3   
5690  assese, off lagos-ibadan expresswayobafemi owo...   250000     2      2   

                                                  extra  state locality  lga  \
53    lovely new & very spacious executive 2 bedroom...  Lagos      NaN  NaN   
217   furnished serviced newly builtannual rent:-350...  Lagos      NaN  NaN   
229   serviced newly builtrent = 1.1m agency = 10% l...  Lagos      NaN  NaN   
324   lovely newly built 2bedroom flat at owode onir...  Lagos      NaN  NaN   
759   just out for letting: massive 3 bedroom apartm...  Lagos      NaN  NaN   
854   with close proximity to the lagos state univer...  Lagos      NaN  NaN   
1079  with close proximity to the lagos state univer...  Lagos      NaN  NaN   
1267  beautifully finished 3 bedroom flat with 4 toi...  Lagos      NaN  NaN   
2199  lovely apartment ensuit 2bedroom 350k per annu...  Lagos      NaN  NaN   
2459  lovely 2bedroom at irawo owode onirin 500k per...  Lagos      NaN  NaN   
2466  a lovely roomself contain apartment with wardr...  Lagos      NaN  NaN   
2722  furnisheda lovely mini flat apartment with war...  Lagos      NaN  NaN   
2981  a brand new and newly built miniflat tolet at ...  Lagos      NaN  NaN   
3048  a newly built 2 bedroom flat wuth luxury finis...  Lagos      NaN  NaN   
3294  s

I'll change the `lga` of the hoods with **Alimosho** and **Ojo** to their `hood`'s name, since they are both local governments in Lagos state.

In [13]:
data.loc[(data.lga.isna()) & (data.hood.str.fullmatch('Ojo') | data.hood.str.fullmatch('Alimosho')), 'lga'] = data.hood

In [14]:
data = data.drop(data[(data.locality.isna()) & (data.lga.isna())].index)
data.shape

(5809, 9)

### Cleaning listings with `LGA` but no `locality`

In [15]:
data[data.locality.isna()]

neighborhood    price  beds  baths  \
6                                  sangotedo ajah lagos   550000     1      1   
17                                    ibeju lekki lagos   350000     1      1   
31                                    ibeju lekki lagos   700000     3      3   
34                                 ...ibeju lekki lagos   700000     1      1   
35                                 ...ibeju lekki lagos   900000     2      2   
36                                 sangotedo ajah lagos  1600000     3      2   
37                                 ...ibeju lekki lagos   600000     1      1   
269                  alafia estateigbogbo ikorodu lagos   200000     2      2   
350        private estate, opicisheri north ojodu lagos   650000     4      3   
352    private estate near opicisheri north ojodu lagos   650000     4      3   
381                                   ikate lekki lagos  2000000     2      2   
386                                   ikate lekki lagos  2000000     2      2   
388                                   ikate lekki lagos  2500000     5      2   
389                                    ikatelekki lagos  2000000     3      3   
399                                   ikate lekki lagos  2500000     3      3   
515                                sangotedo ajah lagos   300000     1      1   
528                                   ibeju lekki lagos   350000     1      1   
557                                   ikate lekki lagos  2300000     2      2   
565               alpha beach roadigbo efon lekki lagos  1200000     2      1   
611                                   ikate lekki lagos  2200000     2      2   
854             iyanera - ketu axisokokomaiko ojo lagos   200000     2      1   
1079                 iyanera - ketuokokomaiko ojo lagos   200000     3      3   
1116                            ...sangotedo ajah lagos   450000     5      2   
1166                                  ikate lekki lagos  2000000     1      2   
1220                       ikate lekkiikate lekki lagos  1200000     1      1   
1267                                     alimosho lagos   350000     4      3   
1499                             opic estateojodu lagos  1000000     2      2   
1574                               sangotedo ajah lagos   800000     2      2   
1634                               sangotedo ajah lagos  1100000     3      3   
1716         erah street ijanikin oko afo badagry lagos   350000     3      3   
1749                  odudu avenueigbogbo ikorodu lagos   200000     3      3   
1818                               sangotedo ajah lagos  2500000     4      4   
1874                               sangotedo ajah lagos  1100000     2      2   
2074                                  ikate lekki lagos  2300000     5      3   
2081                               sangotedo ajah lagos  1200000     3      3   
2144                                  ikate lekki lagos  2300000     3      3   
2212                                  ikate lekki lagos  2300000     2      2   
2649                                  ikate lekki lagos  2200000     2      3   
2659                               sangotedo ajah lagos  1500000     3      3   
2676                               sangotedo ajah lagos   750000     2      2   
2823                               sangotedo ajah lagos   850000     5      3   
2826                                  ikate lekki lagos  2500000     1      2   
2852                            ...sangotedo ajah lagos   800000     2      3   
2866                               sangotedo ajah lagos  2000000     4      5   
2868                               sangotedo ajah lagos   800000     3      4   
2869                               sangotedo ajah lagos   600000     1      2   
2872                               sangotedo ajah lagos  1500000     4      4   
2873                               sangotedo ajah lagos   800000     1      2   
2874                               sangotedo ajah lagos   800000     3    

If **Lekki** is in the listing's `hood` or `lga` (as in Ibeju Lekki), the locality should be **Lekki**.

If the `hood` contains **Ajah**, the locality should be **Ajah**.

If the `lga` is **Obio/Akpor**, the locality is **Port Harcourt**.

If the state is **Oyo** or **Abuja**, the locality should be **Ibadan** and **Abuja** respectively.

In [16]:
data.loc[(data.locality.isna()) & (data.lga.str.contains('Lekki') | data.hood.str.contains('Lekki')), 'locality'] = 'Lekki'
data.loc[(data.locality.isna()) & (data.hood.str.contains('Ajah')), 'locality'] = 'Ajah'
data.loc[(data.locality.isna()) & (data.neighborhood.str.contains('igbogbo')), 'locality'] = 'Igbogbo'

data.loc[(data.locality.isna()) & (data.lga.str.contains('Obio/Akpor')), 'locality'] = 'Port Harcourt'
data.loc[(data.locality.isna()) & (data.state.str.contains('Oyo')), 'locality'] = 'Ibadan'
data.loc[(data.locality.isna()) & (data.state.str.contains('Abuja')), 'locality'] = 'Abuja'

data[data.locality.isna()]

neighborhood    price  beds  baths  \
350        private estate, opicisheri north ojodu lagos   650000     4      3   
352    private estate near opicisheri north ojodu lagos   650000     4      3   
565               alpha beach roadigbo efon lekki lagos  1200000     2      1   
854             iyanera - ketu axisokokomaiko ojo lagos   200000     2      1   
1079                 iyanera - ketuokokomaiko ojo lagos   200000     3      3   
1267                                     alimosho lagos   350000     4      3   
1499                             opic estateojodu lagos  1000000     2      2   
1716         erah street ijanikin oko afo badagry lagos   350000     3      3   
2981                             -shasha alimosho lagos   400000     1      1   
3089                                -amuwo odofin lagos   800000     2      2   
3789  ishasi (ogun state side of ojodu berger)berger...   400000     4      3   
5598    off berger expresswaymagboro obafemi owode ogun   350000     4      2   
5599                                         arepo ogun   500000     4      2   
5600                  oke-afamagboro obafemi owode ogun   600000     3      3   
5602                                   arepo arepo ogun   600000     3      3   
5637  very close to magboro bus stopmagboro obafemi ...   450000     2      2   
5639     27, ogoluwa street ntabo sangoado odo ota ogun   400000     1      1   
5641  ire akari cda,off coca-cola road ojuoreado odo...   200000     2      2   
5642                                       - arepo ogun   600000     2      3   
5647                                      ...arepo ogun   500000     3      3   
5650                               areparepo arepo ogun   650000     3      3   
5659               arepo private estatearepo arepo ogun   300000     2      2   
5663               arepo private estatearepo arepo ogun   650000     3      3   
5665                                         arepo ogun   500000     4      3   
5675              arepo private estate arepo arepo ogun   800000     3      3   
5685             citiview estate, arepoarepo arepo ogun  1500000     3      3   
5688  ishasi (ogun state side of ojodu berger)berger...   400000     4      3   
5693                forthright. gardensarepo arepo ogun   700000     2      2   
5703                     secured estatearepo arepo ogun   500000     4      3   
5704      peace garden estatemagboro obafemi owode ogun   450000     4      2   
5711                   citiview estate, arepoarepo ogun  1500000     3      3   
5712              arepo private estate arepo arepo ogun   600000     3      3   
5713               arepo private estatearepo arepo ogun   450000     2      2   
5720               arepo private estatearepo arepo ogun   600000     3      3   
5724  forthright estate behind punch newspapermagbor...  1000000     3      2   
5728  forthright garden magboro behind punch newspap...  1000000     3      2   
5730                           atan otaado odo ota ogun   150000     1      1   
5736               arepo private estatearepo arepo ogun   500000     2      3   
5742                                ...arepo arepo ogun   600000     3      3   
5751  davol bus stop, iyana-iyesi, otaota idiroko ro...   300000     3      2   
5755             toyin closet kajola obafemi owode ogun   210000     3      3   
5761            58 vincent streetijebu north ijebu ogun   250000     2      1   
5762   onibukunota idiroko road tomori ado odo ota ogun   120000     1      2   

                                                  extra  state locality  \
350   24-hours security spacious rooms water heater ...  Lagos      NaN   
352   24-hours security spacious rooms water heater ...  Lagos      NaN   
565     a decent two bedroom at atlantic view estate..   Lagos      NaN   
854   with close proximity to the lagos state univer...  Lagos      NaN   
1079  with close proximity to the lagos state univer...  Lagos      NaN   
1267  beautifully finished 3

### The Special Cases of Arepo and Magboro

Magboro and Arepo are two towns located beside each other in Obafemi Owode local government of Ogun state (at the suburbs of Lagos state), often confused for each other. Even our Google Maps API got them mixed up!

In [17]:
data.loc[data.neighborhood.str.contains('arepo|magboro')]

neighborhood    price  beds  baths  \
348         private estate near arepoberger ojodu lagos   550000     3      3   
351                  arepo estate viaberger ojodu lagos   500000     4      2   
2254  magboro by mfm prayer city side, off lagos iba...   450000     4      2   
2800                           arepo berger ojodu lagos   450000     2      2   
3282                 arepo estate viaberger ojodu lagos   520000     2      2   
3310                 arepo estate viaberger ojodu lagos   700000     3      3   
3675                           ojodumagboro ojodu lagos  2000000     3      3   
5589           private estatemagboro obafemi owode ogun   250000     1      1   
5590                  magboromagboro obafemi owode ogun   350000     4      2   
5591                           private estatearepo ogun   450000     4      2   
5592  private estate, off berger expresswaymagboro o...   400000     4      3   
5593                    private estateberger arepo ogun   500000     4      2   
5594                      havana estatearepo arepo ogun   500000     4      3   
5595                           private estatearepo ogun   470000     4      2   
5596                           private estatearepo ogun   400000     4      2   
5597                  journalist estatearepo arepo ogun   350000     2      2   
5598    off berger expresswaymagboro obafemi owode ogun   350000     4      2   
5599                                         arepo ogun   500000     4      2   
5600                  oke-afamagboro obafemi owode ogun   600000     3      3   
5602                                   arepo arepo ogun   600000     3      3   
5621  magboro by mfm prayer city side, off lagos iba...   450000     4      2   
5636           private estatemagboro obafemi owode ogun   450000     2      2   
5637  very close to magboro bus stopmagboro obafemi ...   450000     2      2   
5640                            arepoobafemi owode ogun   800000     3      4   
5642                                       - arepo ogun   600000     2      3   
5647                                      ...arepo ogun   500000     3      3   
5648    private estate, via berger expresswayarepo ogun   450000     4      2   
5649                                via arepoarepo ogun   500000     3      3   
5650                               areparepo arepo ogun   650000     3      3   
5651                  magboromagboro obafemi owode ogun   450000     3      3   
5652                  magboromagboro obafemi owode ogun   450000     3      3   
5653                           private estatearepo ogun   600000     4      3   
5654                  magboromagboro obafemi owode ogun   450000     3      3   
5655                 magboro magboro obafemi owode ogun   350000     1      1   
5656                  magboromagboro obafemi owode ogun   220000     1      1   
5657                  magboromagboro obafemi owode ogun   300000     1      2   
5659               arepo private estatearepo arepo ogun   300000     2      2   
5660                  arepo estate roadarepo arepo ogun   550000     3      3   
5663               arepo private estatearepo arepo ogun   650000     3      3   
5665                                         arepo ogun   500000     4      3   
5666                               forthrightarepo ogun   500000     2      2   
5667                    gated estate inarepo arepo ogun   500000     4      2   
5668                  magboromagboro obafemi owode ogun   350000     2      2   
5669                  magboromagboro obafemi owode ogun   500000     2      2   
5673                           private estatearepo ogun   500000     4      2   
5674    private estate, via berger expresswayarepo ogun   800000     3      3   
5675              arepo private estate arepo arepo ogun   800000     3      3   
5676                  magboromagboro obafemi owode ogun   450000     2      2   
5677                  magboromagboro obafemi owode ogun   450000     2    

In [18]:
data.loc[data.neighborhood.str.contains('arepo'), 'locality'] = 'Arepo'
data.loc[data.neighborhood.str.contains('magboro'), 'locality'] = 'Magboro'

# Changing the lga
data.loc[data.locality.str.contains('Magboro|Arepo') == True, 'lga'] = 'Obafemi Owode'

# Changing the state
data.loc[data.locality.str.contains('Magboro|Arepo') == True, 'state'] = 'Ogun'

### The Special Cases of `Ifo` an `Ado Odo/Ota`

Ifo local government in Ogun state is located beside Ado Odo/Ota local governments, at the suburbs of Lagos.

The listings here that are located at Ifo LGA are those at directly beside Ojodu Berger in Lagos, so we'll categorise all as Lagos

In [19]:
data[data.locality.isna()]

neighborhood    price  beds  baths  \
350        private estate, opicisheri north ojodu lagos   650000     4      3   
352    private estate near opicisheri north ojodu lagos   650000     4      3   
565               alpha beach roadigbo efon lekki lagos  1200000     2      1   
854             iyanera - ketu axisokokomaiko ojo lagos   200000     2      1   
1079                 iyanera - ketuokokomaiko ojo lagos   200000     3      3   
1267                                     alimosho lagos   350000     4      3   
1499                             opic estateojodu lagos  1000000     2      2   
1716         erah street ijanikin oko afo badagry lagos   350000     3      3   
2981                             -shasha alimosho lagos   400000     1      1   
3089                                -amuwo odofin lagos   800000     2      2   
3789  ishasi (ogun state side of ojodu berger)berger...   400000     4      3   
5639     27, ogoluwa street ntabo sangoado odo ota ogun   400000     1      1   
5641  ire akari cda,off coca-cola road ojuoreado odo...   200000     2      2   
5688  ishasi (ogun state side of ojodu berger)berger...   400000     4      3   
5730                           atan otaado odo ota ogun   150000     1      1   
5751  davol bus stop, iyana-iyesi, otaota idiroko ro...   300000     3      2   
5755             toyin closet kajola obafemi owode ogun   210000     3      3   
5761            58 vincent streetijebu north ijebu ogun   250000     2      1   
5762   onibukunota idiroko road tomori ado odo ota ogun   120000     1      2   

                                                  extra  state locality  \
350   24-hours security spacious rooms water heater ...  Lagos      NaN   
352   24-hours security spacious rooms water heater ...  Lagos      NaN   
565     a decent two bedroom at atlantic view estate..   Lagos      NaN   
854   with close proximity to the lagos state univer...  Lagos      NaN   
1079  with close proximity to the lagos state univer...  Lagos      NaN   
1267  beautifully finished 3 bedroom flat with 4 toi...  Lagos      NaN   
1499  brand new 2bedroom flat in opic estate for n1m...  Lagos      NaN   
1716  3 bedroom shared apartment flat / apartment fo...  Lagos      NaN   
2981  a brand new and newly built miniflat tolet at ...  Lagos      NaN   
3089  newly built 2 bedroom flat at amuwo odofin nea...  Lagos      NaN   
3789  cool spacious prepaid metre all rooms ensuite ...  Lagos      NaN   
5639  a standard room self for rent with all applian...   Ogun      NaN   
5641         well built 2 bedroom flat neatly tile. ..    Ogun      NaN   
5688  cool spacious prepaid metre all rooms ensuite ...   Ogun      NaN   
5730  newly builtmini flat for rent one room, one to...   Ogun      NaN   
5751  furnishedstandard 3 bedroom flat for rent at d...   Ogun      NaN   
5755  newly builtvery clean 3bedroom flat available ...   Ogun      NaN   
5761                          servicedenough spacing..    Ogun      NaN   
5762  newly builta newly built property with blocks ...   Ogun      NaN   

                lga        hood  
350             Ifo  Isheri Oke  
352             Ifo  Isheri Oke  
565         Eti Osa     Eti-Osa  
854             Ojo         Ojo  
1079            Ojo         Ojo  
1267       Alimosho    Alimosho  
1499            Ifo  Isheri Oke  
1716        Badagry         NaN  
2981       Alimosho    Alimosho  
3089   Amuwo Odofin         NaN  
3789            Ifo         Iju  
5639            Ifo         NaN  
5641    Ado Odo/Ota         NaN  
5688            Ifo         Iju  
5730    Ado Odo/Ota        Atan  
5751    Ado Odo/Ota         NaN  
5755  Obafemi Owode      Kajola  
5761    Ijebu North         NaN  
5762    Ado Odo/Ota         NaN

In [25]:
data[(data.lga.str.contains('Ifo')==True) & (data.hood), 'locality'] = 'Ojodu Berger'
data[(data.lga.str.contains('Ifo')==True) & (data.hood), 'state'] = ''

neighborhood    price  beds  baths  \
101                  ...river valley estate ojodu lagos   600000     1      1   
350        private estate, opicisheri north ojodu lagos   650000     4      3   
352    private estate near opicisheri north ojodu lagos   650000     4      3   
490           island heritage estate berger ojodu lagos   550000     3      3   
491              island heritage estate areaojodu lagos   500000     3      3   
641                           estate berger ojodu lagos  1500000     3      3   
655       island heritage estate areaberger ojodu lagos   900000     3      3   
664   heritage estate ojodu abiodun axis (ogun state...   500000     3      3   
901                                           iju lagos   300000     1      1   
1489                    river valley estate ojodu lagos  1000000     2      2   
1499                             opic estateojodu lagos  1000000     2      2   
1543                        yakoyo alagbole ojodu lagos   250000     1      1   
1676                      just right berger ojodu lagos   700000     2      2   
1841            magodo phase 1omole phase 1 ojodu lagos  1200000     3      3   
1964               alagbole yakoyo alagbole ojodu lagos   480000     2      2   
1965                        yakoyo alagbole ojodu lagos   500000     2      2   
2023             harmony estate iyana ipaja ipaja lagos   700000     3      4   
3268                 mtg gardenisheri north ojodu lagos   900000     2      2   
3296   magodo gra ph.1,isheri........berger ojodu lagos  1300000     3      4   
3320         river valley estate.....berger ojodu lagos   850000     2      3   
3330    beside river vaiiey estate,,,berger ojodu lagos   500000     1      2   
3349         river valley estate.....berger ojodu lagos  1200000     3      4   
3359    beside river vaiiey estate,,,berger ojodu lagos   750000     3      3   
3550        wakati-adurariver valley estate ojodu lagos  1500000     3      3   
3681  off palace road, off westerner avenue berger o...   800000     1      1   
3684                          estate berger ojodu lagos  1500000     3      3   
3726                              off iju roadiju lagos   700000     5      3   
3733          island heritage estate berger ojodu lagos   550000     3      3   
3780               g00dwiii estate,,,berger ojodu lagos   350000     1      1   
3789  ishasi (ogun state side of ojodu berger)berger...   400000     4      3   
3813        estate driveriver valley estate ojodu lagos  1100000     2      3   
3862           island heritage estateberger ojodu lagos   450000     2      2   
4147                 heritage estate berger ojodu lagos   450000     3      3   
4197   magodo gra ph.1,isheri........berger ojodu lagos  1100000     3      3   
4257                 mtr estateisheri north ojodu lagos   900000     2      2   
4302           island heritage estateberger ojodu lagos   450000     2      2   
4312       river valley estate.......berger ojodu lagos  1200000     3      4   
4327                          estate berger ojodu lagos   700000     2      3   
5601  heritage estate ojodu abiodun axis (ogun state...   500000     3      3   
5688  ishasi (ogun state side of ojodu berger)berger...   400000     4      3   

                                                  extra  state      locality  \
101   very nice mini flat @ river valley estate ojod...  Lagos         Ojodu   
350   24-hours security spacious rooms water heater ...  Lagos           NaN   
352   24-hours security spacious rooms water heater ...  Lagos           NaN   
490   secure environment horizon view fresh air cros...  Lagos  Ojodu Berger   
491   cool and spacious 3 bedroom apartment with pop...  Lagos  Ojodu Berger   
641   secure estate. upstairs serene environment sec...  Lagos         Ojodu   
655   newly builtnewly built cool breeze spacious be...  Lagos  Ojodu Berger   
664   neat pay and pack in spacious spacious compoun...  Lagos         Lagos   
9